oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/authentication/OAuth2TokenEndpointAuthenticationSuccessHandler.java

```
@@ -0,0 +1,104 @@
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
+import org.springframework.security.oauth2.core.http.converter.OAuth2AccessTokenResponseHttpMessageConverter;
+import org.springframework.security.oauth2.server.authorization.authentication.OAuth2AccessTokenAuthenticationToken;
+import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An implementation of an {@link AuthenticationSuccessHandler} used for handling an {@link OAuth2AccessTokenAuthenticationToken}
+ * and returning the {@link OAuth2AccessTokenResponse Access Token Response}.
+ *
+ * @author Dmitriy Dubson
+ * @see AuthenticationSuccessHandler
+ * @see OAuth2AccessTokenResponseHttpMessageConverter
+ * @since 1.2
+ */
+public class OAuth2TokenEndpointAuthenticationSuccessHandler implements AuthenticationSuccessHandler {
+	private final Log logger = LogFactory.getLog(getClass());
+
+	private HttpMessageConverter<OAuth2AccessTokenResponse> accessTokenHttpResponseConverter =
+			new OAuth2AccessTokenResponseHttpMessageConverter();
+
+	@Override
+	public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authentication) throws IOException, ServletException {
+		AuthenticationSuccessHandler.super.onAuthenticationSuccess(request, response, chain, authentication);
+	}
+
+	@Override
+	public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication) throws IOException, ServletException {
+		if (!(authentication instanceof OAuth2AccessTokenAuthenticationToken accessTokenAuthentication)) {
+			// TODO: determine what should the HTTP response be if authentication type is incorrect.
+			if (this.logger.isWarnEnabled()) {
```

The `authentication` must be of type `OAuth2AccessTokenAuthenticationToken`, otherwise, this `AuthenticationSuccessHandler` is misconfigured. We should throw an exception in these cases.

---

Gotcha. I've updated the block with an exception thrown, however I'm not 100% on the wording of the error message itself. That may need some polishing.